include/RED4ext/SharedPtr.hpp

```diff
@@ -0,0 +1,377 @@
+            RelocFunc<void (*)(SharedPtrBase*)> decWeakRefFn(Addresses::Handle_DecWeakRef);
+            decWeakRefFn(this);
+        }
+    }
+};
+RED4EXT_ASSERT_SIZE(SharedPtrBase<void>, 0x10);
+RED4EXT_ASSERT_OFFSET(SharedPtrBase<void>, instance, 0x00);
+RED4EXT_ASSERT_OFFSET(SharedPtrBase<void>, refCount, 0x08);
+
+template<typename T>
+class SharedPtrWithAccess : public SharedPtrBase<T>
+{
+public:
+    template<typename U = T, typename = std::enable_if_t<!std::is_void_v<U>>>
+    [[nodiscard]] inline U& operator*() const
+    {
+        return *GetPtr();
+    }
+
+    [[nodiscard]] inline T* operator->() const
+    {
+        return GetPtr();
+    }
+
+    [[nodiscard]] inline operator T*() const
+    {
+        return GetPtr();
+    }
+
+    explicit operator bool() const noexcept
+    {
+        return GetPtr() != nullptr;
+    }
+
+    [[nodiscard]] T* GetPtr() const noexcept
+    {
+        return reinterpret_cast<T*>(this->instance);
+    }
+};
+
+template<typename T>
+class WeakPtrWithAccess : public SharedPtrBase<T>
+{
+public:
+    explicit operator bool() const noexcept
+    {
+        return !Expired();
+    }
+
+    [[nodiscard]] bool Expired() const noexcept
+    {
+        return this->GetUseCount() == 0;
+    }
+};
+
+template<typename T>
+class WeakPtr;
+
+template<typename T>
+class SharedPtr : public SharedPtrWithAccess<T>
+{
+public:
+    constexpr SharedPtr() noexcept = default;
+    constexpr SharedPtr(std::nullptr_t) noexcept
+    {
+    }
+
+    explicit SharedPtr(T* aPtr)
+    {
+        this->instance = aPtr;
+        this->refCount = Memory::New<RefCnt>();
+    }
+
+    SharedPtr(const SharedPtr& aOther) noexcept
+    {
+        this->CopyConstructFrom(aOther);
+    }
+
+    SharedPtr(SharedPtr&& aRhs) noexcept
+    {
+        this->MoveConstructFrom(std::move(aRhs));
+    }
+
+    SharedPtr(const WeakPtr<T>& aOther)
+    {
+        this->TryCopyConstructFromWeak(aOther);
+    }
+
+    ~SharedPtr()
+    {
+        static_assert(Memory::detail::HasAnyAllocator<T> && Memory::detail::IsDestructible<T>,
+                      "SharedPtr only supports types that define the allocator type and are destructible "
+                      "(a polymorphic type requires a virtual destructor).");
+
+        if (this->DecRef())
+        {
+            this->Destroy();
+        }
+    }
+
+    SharedPtr& operator=(const SharedPtr& aRhs) noexcept
+    {
+        SharedPtr(aRhs).Swap(*this);
+        return *this;
+    }
+
+    SharedPtr& operator=(SharedPtr&& aRhs) noexcept
+    {
+        SharedPtr(std::move(aRhs)).Swap(*this);
+        return *this;
+    }
+
+    void Reset() noexcept
+    {
+        SharedPtr().Swap(*this);
+    }
+
+    void Reset(T* aPtr)
+    {
+        SharedPtr(aPtr).Swap(*this);
+    }
+
+    void Swap(SharedPtr& aOther) noexcept
+    {
+        this->DoSwap(aOther);
+    }
+};
+RED4EXT_ASSERT_SIZE(SharedPtr<void*>, 0x10);
+
+template<typename T>
+class WeakPtr : public WeakPtrWithAccess<T>
```

Comment:
Can you move this class (and its friends) to its own file?

Reply:
Include it in `SharedPtr`? 
What about `WeakHandle`?

Reply:
>Include it in SharedPtr?
I would say just forward declare it.

> What about WeakHandle?
Include it there, I would also move it to its own file.

Reply:
Current code including generated natives expect `WeakHandle` to be available when `Handle.hpp` is included.
We can switch to separate headers but include weak in strong for compatibility. Wdyt?

Reply:
What if you only forward declare `WeakHandle` in `SharedPtr` and include `SharedPtr` in `WeakHandle`? Wouldn't that be better or I missed some relationship?

Reply:
I want to avoid defining `SharedPtr` even when it's not needed.
Like for example `WeakHandle` only needs base class `SharedPtrBase`, but base class is in the same header with `SharedPtr`. So with `WeakHandle` you always include `SharedPtr`. Same with `Handle`, even though it does't need `SharedPtr`.

Reply:
Then let's let it like this.